# Louisville Free Public Library

Analysis of Louisville Free Public Library collection inventory.

Data Source: [Louisville Metro Open Data](https://data.louisvilleky.gov/datasets/LOJIC::louisville-metro-ky-library-collection-inventory-/about)

# LFPL's collection inventory. Updated on a monthly basis.

## Definitions:
- **BibNum** - The unique identifier of a bibliographic record within our materials database. Materials with the same bibliographic # will generally have the same cataloging metadata, differing only in the barcode number, assigned location and anything else specific to the individual copy.

- **Title** - The name of the material.

- **Author** - The writer or creator of the material.

- **ISBN** - The International Standard Book Number is a numeric commercial book identifier that is intended to be unique. Publishers purchase ISBNs from an affiliate of the International ISBN Agency. An ISBN is assigned to each separate edition and variation of a publication. 

- **Publication Year**- The year that the material was originally published.

- **Item Type** - Describes the type of material of each item, including Books, Audiobooks, Serials, DVDs, Microforms, Three Dimensional Objects, Kits, and Printed Cartographic Materials. 

- **Item Collection** - Refers to the collection the material belongs to based on common themes, including but not limited to Adult Fiction, Adult Reference, Mystery, Children’s Fiction, etc.  

- **Item Location** - The library location where the material was assigned at the time the report was run.

- **Item Price** – The price, in USD, that LFPL purchased the material for. 

***
***

### 1. Following the Demo & Exploring the Data
***

In [5]:
#load in the data
import pandas as pd
books_df = pd.read_csv('books.csv.gz')

In [6]:
#exploring the columns
books_df.columns

Index(['BibNum', 'Title', 'Author', 'ISBN', 'PublicationYear', 'ItemType',
       'ItemCollection', 'ItemLocation', 'ItemPrice', 'ReportDate'],
      dtype='object')

In [7]:
#rows and columns count
books_df.shape

(1190176, 10)

In [8]:
#getting the data types from columns
books_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1190176 entries, 0 to 1190175
Data columns (total 10 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   BibNum           1190176 non-null  int64  
 1   Title            1190175 non-null  object 
 2   Author           1124225 non-null  object 
 3   ISBN             1153891 non-null  float64
 4   PublicationYear  1190176 non-null  int64  
 5   ItemType         1190176 non-null  object 
 6   ItemCollection   1190036 non-null  object 
 7   ItemLocation     1190176 non-null  object 
 8   ItemPrice        1190176 non-null  float64
 9   ReportDate       1190176 non-null  object 
dtypes: float64(2), int64(2), object(6)
memory usage: 90.8+ MB


In [10]:
#getting missing record count
books_df.isnull().sum().sort_values(ascending=False)

Author             65951
ISBN               36285
ItemCollection       140
Title                  1
BibNum                 0
PublicationYear        0
ItemType               0
ItemLocation           0
ItemPrice              0
ReportDate             0
dtype: int64

In [11]:
#previewing the data
books_df.head()


BibNum                                              Title  \
0  707409  Jeff Immelt and the new GE way : innovation, t...   
1  707411  Robin rescues dinner : 52 weeks of quick-fix m...   
2  707411  Robin rescues dinner : 52 weeks of quick-fix m...   
3  707411  Robin rescues dinner : 52 weeks of quick-fix m...   
4  707411  Robin rescues dinner : 52 weeks of quick-fix m...   

                 Author          ISBN  PublicationYear ItemType  \
0   Magee, David, 1965-  9.780072e+12             2009     Book   
1  Miller, Robin, 1964-  9.780307e+12             2009     Book   
2  Miller, Robin, 1964-  9.780307e+12             2009     Book   
3  Miller, Robin, 1964-  9.780307e+12             2009     Book   
4  Miller, Robin, 1964-  9.780307e+12             2009     Book   

      ItemCollection            ItemLocation  ItemPrice           ReportDate  
0  Adult Non-Fiction                    Main      25.95  02/01/2023 00:00:00  
1  Adult Non-Fiction               Southwest      19.99  02/01/2023 00:00:00  
2  Adult Non-Fiction               Southwest      19.99  02/01/2023 00:00:00  
3  Adult Non-Fiction  Remote Shelving - Main      19.99  02/01/2023 00:00:00  
4  Adult Non-Fiction  Remote Shelving - Main      19.99  02/01/2023 00:00:00

In [16]:
#exploring the Item Type column
books_df['ItemType'].unique()
books_df['ItemType'].value_counts()

Book    1190176
Name: ItemType, dtype: int64

In [18]:
#exploring the Item Collection column
books_df['ItemCollection'].unique()
books_df['ItemCollection'].value_counts()

Adult Non-Fiction                   371433
Adult Fiction                       177604
Children's Non-Fiction               86356
Mystery                              60314
Children's Picture Book              59348
Preschool  Picture Book              51276
Children's Fiction                   48446
Adult Paperback                      45302
Children's Paperback                 45076
Children's Easy Reader               24511
Teen Non-Fiction                     24376
Older Teen Fiction                   23787
Children's Board Book                20057
Younger Teen  Fiction                17532
Kentucky History                     16962
Science Fiction                      16048
Children's Easy Reader Paperback     15959
Holiday                              15583
International Collection             15581
Adult Reference                      11197
Children's Picture Paperback          9731
Urban Fiction                         7601
Caldecott/Newbery                     6777
Storytime C

In [25]:
#eploring the ItemLocation column
books_df['ItemLocation'].unique()
books_df['ItemLocation'].value_counts()

Remote Shelving - Main       139987
Northeast                    124473
Southwest                    122113
Main                         121439
South Central                115837
Bon Air                       74730
St Matthews                   69531
Jeffersontown                 56706
Iroquois                      52382
Highlands - Shelby Park       45539
Crescent Hill                 42837
Childrens Main Library        38994
Middletown                    33120
Shively                       23623
Newburg                       23586
Fairdale                      23149
Shawnee                       22906
Western                       21648
Portland                      13334
Childrens Bookmobile           9129
Remote Shelving - Shawnee      9083
Main Teen                      6024
Content Management                4
Adult Bookmobile                  2
Name: ItemLocation, dtype: int64

In [23]:
#eploring the Author column
books_df['Author'].unique()
books_df['Author'].value_counts()

Patterson, James, 1947-                 5856
Osborne, Mary Pope.                     2063
Steel, Danielle                         1824
Pilkey, Dav, 1966-                      1812
Seuss, Dr.                              1812
                                        ... 
Sánchez Ferlosio, Rafael, 1927-2019       1
Pippen, Kitty, 1919-2018                   1
Adkins, Frank (Francis A.)                 1
Ray, James A.                              1
Kniffke, Sophie.                           1
Name: Author, Length: 187472, dtype: int64

In [29]:
#eploring item price values

result = books_df['ItemPrice'].describe()
result = result.apply(lambda x: format(x, ',.2f'))
print(result)

count    1,190,176.00
mean            18.45
std             16.00
min              0.00
25%             10.95
50%             15.99
75%             24.95
max          1,077.00
Name: ItemPrice, dtype: object


In [34]:
#eploring publication year values

result = books_df['PublicationYear'].describe()
result = result.apply(lambda x: format(int(x), 'd'))
print(result)

count    1190176
mean        2004
std          101
min            0
25%         2005
50%         2014
75%         2018
max         9999
Name: PublicationYear, dtype: object


### 2. A Little Exploration Tangent with Publication Year
***

**Publication Year** in this instance is categorical. We are not using it to measure time between two years, instead it is a classification desiginated to the book. It would be interesting to see the count of years within the records instead, or a range of years that the books may fall into. Let's do that instead. 

In [41]:
#exploring Publication Year
books_df['PublicationYear'].unique()
books_df['PublicationYear'].value_counts()

2018    77168
2017    72501
2015    70429
2016    70174
2020    68074
        ...  
1818        1
1794        1
1814        1
1807        1
1829        1
Name: PublicationYear, Length: 213, dtype: int64

This is too much data to digest in a meaningful way so I wonder if I can break it into groupings

In [45]:
bins = [1700, 1800, 1900, 2000, 2100]
labels = ['1700-1799', '1800-1899', '1900-1999', 'Present day']
books_df['PublicationCategory'] = pd.cut(books_df['PublicationYear'], bins=bins, labels=labels, right=False)
result = books_df['PublicationCategory'].value_counts()
print(result)

Present day    1013675
1900-1999       172198
1800-1899         1323
1700-1799            3
Name: PublicationCategory, dtype: int64


I'm still new to this, so I want to double check that this makes sense with the data set we have so I'll look at the lowest publication dates we have that aren't 0 so make sure

In [51]:
#seeing what values are listed as the lowest 5 that are not 0
results = books_df[books_df['PublicationYear'] !=0]['PublicationYear'].sort_values().head(5)
print(results.values)


[1790 1794 1798 1800 1800]


Doing the above made me realize I likely have 0 and null values present in my bins. I'm not sure how to fix this but I'd like to make note of how many those are

In [58]:
#looking into getting a count of records where PublicationYear is null or 0
resultszero = books_df[books_df['PublicationYear'] ==0]
countzero = resultszero.shape[0]

resultsnan = books_df[books_df['PublicationYear'].isna()]
countnan = resultsnan.shape[0]

print("Record count where publication year is 0 - ", countzero)
print("Record count where publication year is missing - ", countnan)

Record count where publication year is 0 -  2971
Record count where publication year is missing -  0


This tells me there are **2971** records that have 0 listed as a Publication Year. Getting a 0 makes sense for missing values, since our earlier query let us know there are no null values for Publication Year. I don't quite know how to handle that with bins, so I'm leaving it for now

### 3. Returning to the Homework to Explore YA Genre
***

#### How much was spent on the collection for YA? How does that compare to other collections?

In [93]:
#exploring the YA genre

books2_df = books_df[["ItemCollection", "ItemPrice"]]
grouped = books2_df.groupby("ItemCollection").sum()
grouped['ItemPrice'] = grouped['ItemPrice'].apply(lambda x: format((x), ',.2f'))
grouped.sort_values(by='ItemPrice',ascending=False, inplace=False)                                             
print(grouped)


                                     ItemPrice
ItemCollection                                
Adult Audiobook                          39.39
Adult DVD                               149.95
Adult Fiction                     3,457,884.27
Adult Non-Fiction                 8,595,104.56
Adult Paperback                     277,216.10
Adult Paperbacks Tall                 6,239.87
Adult Reference                     615,411.54
Bestsellers                          39,735.84
Big Book                              1,957.34
Book Discussion Kit                      93.30
Caldecott/Newbery                    96,166.58
Children's Board Book               144,366.00
Children's Easy Reader              301,445.43
Children's Easy Reader Paperback     59,366.71
Children's Fiction                  636,529.77
Children's Music                          7.51
Children's Non-Fiction            1,579,063.64
Children's Non-Fiction Paperback        457.68
Children's Paperback                227,002.62
Children's Pi

It looks like compared to children and adult books, there was significantly less money spent on Young Adult novels